Add enhanced Material/Item Color(Type)-API

TheFruxz opened the pull request:
# Material Color API

Adding API to easily manage colors for materials and items

###### Improved version

TheFruxz committed:
Add enhanced Material/Item Color(Type)-API

TheFruxz committed:
Respecting Nulls and NotNulls :)

TheFruxz committed:
More Null and NotNull stuff

Trigary commented on Spigot-API-Patches/0317-Adding-enhanced-Material-Item-Color-Type-API.patch:
```diff
@@ -0,0 +1,594 @@
++    }
++
++    /**
++     * A {@link Set<Material>} of all possible colored {@link Material}-variants of this {@link DyeableMaterial} (material-group)
```
JavaDoc links don't accept generic parameters, this is why the build is failing.

TheFruxz replied:
Thanks for the help!

Trigary reviewed (changes requested):
Marked why the build fails.

TheFruxz committed:
javadoc

TheFruxz committed:
Removed .DS_Store

MiniDigger commented on Spigot-API-Patches/0317-Adding-enhanced-Material-Item-Color-Type-API.patch:
```diff
@@ -0,0 +1,594 @@
++     */
++    @NotNull
++    public Color getParticleColor() {
++        Color result;
```
with java 16 we can use a switch expression here, right?, but even with java 8, the cases should be `return Color.whatever;`, not assigning a var and breaking.

TheFruxz replied:
Yes, it's true, I still had an incorrect JDK loaded, which meant I didn't have the option to do this. I thought I was wrong when I wanted to do it that way, however it was the JDK.
I'm on it ...

TheFruxz committed:
Now using switch'es from Java-14 Level

kashike commented on Spigot-API-Patches/0317-Adding-enhanced-Material-Item-Color-Type-API.patch:
```diff
@@ -0,0 +1,586 @@
++ */
++public enum ColorType {
++
++    NONE,
```
I disagree with this being an option.

TheFruxz replied:
Removed this option entirely

TheFruxz committed:
Removed Non-Color (NONE) state

TheFruxz committed:
Removed DyeableMaterial#getDefaultMaterial function

Removed the getDefaultMaterial function from the DyeableMaterial's, because we don't use it anymore and for the 'public' use it won't fit anymore, because after the 1.12 released, no 'default' colors of materials exists

kezz commented on Spigot-API-Patches/0317-Adding-enhanced-Material-Item-Color-Type-API.patch:
```diff
@@ -0,0 +1,489 @@
++     * @return the {@link ChatColor} of this {@link ColorType}
++     */
++    @NotNull
++    public ChatColor getChatColor() {
```
Personally I don't think that new API should be created that provides access to deprecated stuff.

Proximyst replied:
I concur

kezz commented on Spigot-API-Patches/0317-Adding-enhanced-Material-Item-Color-Type-API.patch:
```diff
@@ -0,0 +1,489 @@
++     */
++    @NotNull
++    public Material getWool() {
++        return Objects.requireNonNull(Material.matchMaterial(this.name() + "_WOOL"));
```
Is there not a smarter way of doing this? Iterating through a tag or something similar?

Proximyst replied:
Constructor & field

kezz commented on Spigot-API-Patches/0317-Adding-enhanced-Material-Item-Color-Type-API.patch:
```diff
@@ -0,0 +1,489 @@
++     */
++    @NotNull
++    public static Set<Material> getAllVariants() {
++        Set<Material> result = new java.util.HashSet<>(Collections.emptySet());
```
These sets should definitely be calculated in a static block or something as this is quite a costly call.

Proximyst replied:
Pre-set size and use static block like Kezz says. Guava ImmutableSet can be used to avoid clones.

kezz commented on Spigot-API-Patches/0317-Adding-enhanced-Material-Item-Color-Type-API.patch:
```diff
@@ -0,0 +1,489 @@
++/**
++ * Represents different kind of possible Colors, which colorable {@link Material}s also/and {@link DyeableMaterial}s can have
++ */
++public enum ColorType {
```
Without a `NONE`, this could perhaps implement `RGBLike`?

kezz reviewed (commented):
Left a few minor thoughts but I'm not sure I'm a massive fan of this as an API addition - it feels slightly over engineered perhaps?

Proximyst commented on Spigot-API-Patches/0317-Adding-enhanced-Material-Item-Color-Type-API.patch:
```diff
@@ -0,0 +1,489 @@
++
++        for (ColorType colorType : values()) {
++            result.addAll(
++                Arrays.asList(
```
`List.of`

Proximyst commented on Spigot-API-Patches/0317-Adding-enhanced-Material-Item-Color-Type-API.patch:
```diff
@@ -0,0 +1,489 @@
++     * @return the {@link ColorType} of the {@link Material} <i>material</i>, or null if it has no supported color
++     */
++    @Nullable
++    public static ColorType getColorType(@NotNull Material material) {
```
Store equivalent in constructor & field

Proximyst commented on Spigot-API-Patches/0317-Adding-enhanced-Material-Item-Color-Type-API.patch:
```diff
@@ -0,0 +1,489 @@
++     */
++    @NotNull
++    public DyeColor getDyeColor() {
++        return DyeColor.valueOf(name());
```
Constructor & field

Proximyst commented on Spigot-API-Patches/0317-Adding-enhanced-Material-Item-Color-Type-API.patch:
```diff
@@ -0,0 +1,489 @@
++    public Color getParticleColor() {
++        return switch (this) {
++
++            case WHITE -> Color.WHITE;
```
Constructor & field

Proximyst commented on Spigot-API-Patches/0317-Adding-enhanced-Material-Item-Color-Type-API.patch:
```diff
@@ -0,0 +1,489 @@
++     */
++    @NotNull
++    public Material getTerracotta() {
++        return Objects.requireNonNull(Material.matchMaterial(this.name() + "_TERRACOTTA"));
```
You know the drill

Proximyst commented on Spigot-API-Patches/0317-Adding-enhanced-Material-Item-Color-Type-API.patch:
```diff
@@ -0,0 +1,489 @@
++
++    @Override
++    public String toString() {
++        return "" + getChatColor();
```
Don't override this.

Proximyst commented on Spigot-API-Patches/0317-Adding-enhanced-Material-Item-Color-Type-API.patch:
```diff
@@ -0,0 +1,489 @@
++     */
++    @NotNull
++    public Material withColor(@NotNull ColorType colorType) {
++        return colorToMaterialProcess.apply(colorType);
```
`this.`

Proximyst commented on Spigot-API-Patches/0317-Adding-enhanced-Material-Item-Color-Type-API.patch:
```diff
@@ -0,0 +1,489 @@
++     */
++    @Nullable
++    public static DyeableMaterial fromMaterial(@NotNull Material material) {
++        for (DyeableMaterial dyeableMaterial : values()) {
```
Use a switch or better yet, tags with abstract method.

Proximyst commented on Spigot-API-Patches/0317-Adding-enhanced-Material-Item-Color-Type-API.patch:
```diff
@@ -0,0 +1,489 @@
++     */
++    @NotNull
++    public Material changeColor(@NotNull ColorType newColor) throws IllegalArgumentException {
++            DyeableMaterial material = getDyeableMaterial();
```
No tabs

Proximyst commented on Spigot-API-Patches/0317-Adding-enhanced-Material-Item-Color-Type-API.patch:
```diff
@@ -0,0 +1,489 @@
++            DyeableMaterial material = getDyeableMaterial();
++            if (material != null) {
++                return material.withColor(newColor);
++            } else
```
New block

Proximyst commented on Spigot-API-Patches/0317-Adding-enhanced-Material-Item-Color-Type-API.patch:
```diff
@@ -0,0 +1,489 @@
++            if (material != null) {
++                return material.withColor(newColor);
++            } else
++                throw new IllegalArgumentException("The material " + this.name() + " does not have color-support!");
```
```suggestion
+                throw new IllegalArgumentException("The material " + this.name() + " is not coloured");
```

Proximyst commented on Spigot-API-Patches/0317-Adding-enhanced-Material-Item-Color-Type-API.patch:
```diff
@@ -0,0 +1,489 @@
++     * @return the {@link ColorType} of this {@link Material}, or <i>other</i> if it has no supported {@link ColorType} color
++     */
++    @NotNull
++    public ColorType getColorTypeOrOther(@NotNull ColorType other) {
```
No. Just let the caller use `Objects#requireNonNullElse` or the likes.

Proximyst commented on Spigot-API-Patches/0317-Adding-enhanced-Material-Item-Color-Type-API.patch:
```diff
@@ -0,0 +1,489 @@
++     * @return the {@link ColorType} of this {@link Material}, or {@link ColorType#WHITE} if it has no supported {@link ColorType} color
++     */
++    @NotNull
++    public ColorType getColorTypeOrWhite() {
```
No

Proximyst commented on Spigot-API-Patches/0317-Adding-enhanced-Material-Item-Color-Type-API.patch:
```diff
@@ -0,0 +1,489 @@
++     * @param newColor the new color, which the {@link Material} will become
++     * @throws IllegalArgumentException if the {@link ItemStack} {@link Material} has no {@link DyeableMaterial}-group
++     */
++    public void changeColor(@NotNull ColorType newColor) throws IllegalArgumentException {
```
Unsure about this; Material provides the method and this doesn't explicitly say it uses `setType`

Proximyst commented on Spigot-API-Patches/0317-Adding-enhanced-Material-Item-Color-Type-API.patch:
```diff
@@ -0,0 +1,489 @@
++     * @throws IllegalArgumentException if the {@link ItemStack} {@link Material} has no {@link DyeableMaterial}-group
++     */
++    public void changeColor(@NotNull ColorType newColor) throws IllegalArgumentException {
++        type = type.changeColor(newColor);
```
`this.setType` in case we need to do verification

Proximyst commented on Spigot-API-Patches/0317-Adding-enhanced-Material-Item-Color-Type-API.patch:
```diff
@@ -0,0 +1,489 @@
++     * @return the color of this {@link ItemStack}s-{@link Material}, or {@link ColorType#WHITE} if it has no supported color
++     */
++    @NotNull
++    public ColorType getColorTypeOrWhite() {
```
No

Proximyst commented on Spigot-API-Patches/0317-Adding-enhanced-Material-Item-Color-Type-API.patch:
```diff
@@ -0,0 +1,489 @@
++     * @return the color of this {@link ItemStack}s-Material, or <i>other</i> if it has no supported color
++     */
++    @NotNull
++    public ColorType getColorTypeOrOther(@NotNull ColorType other) {
```
No

Proximyst commented on Spigot-API-Patches/0317-Adding-enhanced-Material-Item-Color-Type-API.patch:
```diff
@@ -0,0 +1,489 @@
++     */
++    @Nullable
++    public ColorType getColorType() {
++        return ColorType.getColorType(type);
```
```suggestion
+        return this.type.getColorType();
```

TheFruxz committed:
Some changes

TheFruxz committed:
A single change

A248 commented on Spigot-API-Patches/0317-Adding-enhanced-Material-Item-Color-Type-API.patch:
```diff
@@ -0,0 +1,721 @@
++     * Gets all supported colored {@link Material}s, which can be used
++     */
++    @NotNull
++    public static final ImmutableSet<Material> allVariants = ImmutableSet.copyOf(
```
Why not simply return a `Set` here and use the standard `Set.copyOf`? ImmutableSet seems unnecessary and inconsistent to me. (The JDK sets are very nice and have some cool features too)

SirYwell replied:
I wouldn't expose it as a field at all

A248 commented on Spigot-API-Patches/0317-Adding-enhanced-Material-Item-Color-Type-API.patch:
```diff
@@ -0,0 +1,721 @@
++        Set<Material> allVariants = ColorType.allVariants;
++        if (allVariants.stream().anyMatch(it -> it == material)) {
++            for (Material loopMaterial : allVariants) {
++                if (loopMaterial == material) {
```
The implementation could certainly be improved here. You're mixing streams and imperative code, and moreover, you don't need to check twice whether the material you're looking for is present.

A248 commented on Spigot-API-Patches/0317-Adding-enhanced-Material-Item-Color-Type-API.patch:
```diff
@@ -0,0 +1,721 @@
++     * @return all possible colored {@link Material}s of this color
++     */
++    @NotNull
++    public ImmutableSet<Material> getAllMaterials() {
```
Here too, the standard `Set` is much preferable to ImmutableSet.

A248 commented on Spigot-API-Patches/0317-Adding-enhanced-Material-Item-Color-Type-API.patch:
```diff
@@ -0,0 +1,721 @@
++    BANNER_WALL(ColorType::getBannerWall),
++    DYE(ColorType::getDye);
++
++    public Function<ColorType, Material> colorToMaterialProcess;
```
```suggestion
+    private final Function<ColorType, Material> colorToMaterialProcess;
```

A248 commented on Spigot-API-Patches/0317-Adding-enhanced-Material-Item-Color-Type-API.patch:
```diff
@@ -0,0 +1,721 @@
++     */
++    @NotNull
++    public Set<Material> getAllVariants() {
++        return Arrays.stream(ColorType.values()).map(colorToMaterialProcess).collect(Collectors.toSet());
```
```suggestion
+        return Arrays.stream(ColorType.values()).map(colorToMaterialProcess).collect(Collectors.toUnmodifiableSet());
```
I don't suggest using Collectors.toSet() here, for a very particular reason. In fact, I don't suggest *ever* returning the result of Collectors.toSet() from API methods.

It's ambiguous whether the returned Set is immutable or not or whether it accepts nulls. The documentation of explicitly says these qualities are unspecified. Propagating a Set with undefined qualities throughout API return values is error-prone, should a future JDK implementation change the concrete type of the returned Set.

A248 commented on Spigot-API-Patches/0317-Adding-enhanced-Material-Item-Color-Type-API.patch:
```diff
@@ -0,0 +1,721 @@
++     * @param newColor the new color, which the {@link Material} will become
++     * @throws IllegalArgumentException if the {@link ItemStack} {@link Material} has no {@link DyeableMaterial}-group
++     */
++    public void changeTypeColor(@NotNull ColorType newColor) throws IllegalArgumentException {
```
I think it would be better if you left this method to user code. It seems quite awkward to have this here. It'd be clearer to have callers perform this operation.

A248 reviewed (commented):
Like Kezz, I have my reserves about some of these additions. Some of them seem awkward, maybe that's just me.

For now, here are my comments on the specifics of the PR.

SirYwell commented on Spigot-API-Patches/0317-Adding-enhanced-Material-Item-Color-Type-API.patch:
```diff
@@ -0,0 +1,721 @@
++            if (material != null) {
++                return material.withColor(newColor);
++            } else {
++                throw new IllegalArgumentException("The material " + this.name() + " is not coloured!");
```
You're using `color` everywhere else
```suggestion
+                throw new IllegalArgumentException("The material " + this.name() + " is not colored!");
```

SirYwell reviewed (commented):
Not sure if I like that approach in general. Currently, newly added blocks with different color variants need their own new field and method in the `ColorType` enum and renamed blocks cause the same issues we face with the Material enum in general (for both `ColorType` and `DyableMaterial`). While it might be handy to have the appropriate carpet material for a specific wool block, I'd prefer a more tag-like approach.

TheFruxz committed:
Some stuff

TheFruxz committed:
Cleaned up a bit

TheFruxz committed:
not looping every material, only every value
